### ISSUE TITLE
Netcdf writeAttribute NullPointerException

### ISSUE THREAD

**Issue by antoinetran** [Unidata/thredds#925]
In netcdf4-4.6.4 and even latest 4.6.10 (https://github.com/Unidata/thredds/blob/v4.6.10/netcdf4/src/main/java/ucar/nc2/jni/netcdf/Nc4Iosp.java):

Writting an Attribute with not null name and null values creates a NullPointerException at values.getStorage(). In my case:
`ret = nc4.nc_put_att_int(grpid, varid, att.getShortName(), Nc4prototypes.NC_INT, new SizeT(att.getLength()), (int[]) values.getStorage());`

I am not sure a null values is Netcdf compliant, but according to the spec https://www.unidata.ucar.edu/software/netcdf/docs/attribute_conventions.html §"missing values", this is something that can happen. I can even write a null values with Netcdf4 Java with ` attribute = new Attribute(name, dataType, false);`

**Comment by DennisHeimbigner** [comment]
The method nc_put_att_int is not part of the Nc4Iosp
interface. I infer then that you are actually writing to
an instance of Nc4prototypes. This is not a good idea
unless you are an expert. I think using the  method
Nc4Iosp.makeAttribute would be a better option.,

**Comment by DennisHeimbigner** [comment]
The code does not do a null check. The question is: should it just silently return,
or should it generate some other kind of exception?

**Comment by DennisHeimbigner** [comment]
The same bug appears in version 5 also.

**Comment by DennisHeimbigner** [comment]
Ok, I believe the correct solution is that if values == null, then treat it like |values|==0.
|values|==0 should work.

**Comment by antoinetran** [comment]
> Ok, I believe the correct solution is that if values == null, then treat it like |values|==0.
|values|==0 should work.

If values is null, it would be better to make it null at the end.

> The code does not do a null check. The question is: should it just silently return,
or should it generate some other kind of exception?

No, because it is authorized to have null values for an Attribute, if I read correctly the Netcdf spec.

**Comment by DennisHeimbigner** [comment]
> No, because it is authorized to have null values for an Attribute, if I read correctly the Netcdf spec.
I am not sure how you are inferring this. The underlying netcdf-c library and spec has no notion of NULL
value. It does have _FillValue (aka "missing value") but that is a different thing altogether.
Can you elaborate on what you mean by null values.

**Comment by antoinetran** [comment]
> I am not sure how you are inferring this. The underlying netcdf-c library and spec has no notion of NULL
value. It does have _FillValue (aka "missing value") but that is a different thing altogether.
Can you elaborate on what you mean by null values.

Here are the reasons:
1. I am indeed referring to the "missing values" (c https://www.unidata.ucar.edu/software/netcdf/docs/attribute_conventions.html), but I must have misinterpreted "missing values" with null?
2.  In our project, we got a Netcdf file with null values attributes. I'm not sure though if this is a nominal Netcdf file. So I expected after a read and write of the file without altering, that we have the same file. 
3. Moreover, Attribute class allows constructing it without values:
`public Attribute(String name, DataType dataType, boolean isUnsigned)`
and `protected void setValues(Array arr)` is also ok with null.

But if a more export on Netcdf says it is implicitely forbidden to have null values, then I would say it is a business decision to interpret the null, instead of returning 0, which is completely different. For instance when we open the Netcdf, we have to check in our code the validity of this netcdf...

**Comment by antoinetran** [comment]
Up

**Comment by lesserwhirls** [comment]
Do you have a sample file you could share that has an encoded null value?

**Comment by antoinetran** [comment]
[W_xx-YYYYYYYYYY,SAT,SGA1-3MI-1B-RAD_C_YYYY_20220101121212_G_O_20220101103000_20220101104000_C_N____.zip](https://github.com/Unidata/thredds/files/1897747/W_xx-YYYYYYYYYY.SAT.SGA1-3MI-1B-RAD_C_YYYY_20220101121212_G_O_20220101103000_20220101104000_C_N____.zip)

**Comment by antoinetran** [comment]
Hi,

Whatever the policy we decide to do on the Null values, we still need to take care of that so that we avoid a NullPointerException.

**Comment by lesserwhirls** [comment]
Are you reading the netCDF-4 file using the C Library calls directly, and not `ucar.nc2.NetcdfFile`, or `ucar.nc2.dataset.NetcdfDataset`, or something higher up in the netCDF-Java stack?

**Comment by antoinetran** [comment]
@lesserwhirls The netcdf file is read using purely Java Netcdf library. The type of object is indeed the NetcdfFile. We use C library only for writing, because the Java Netcdf library requires that.

**Comment by lesserwhirls** [comment]
Ah, now I see. JNA seems to be happy to pass along a null value to the C library, and the C library seems to be happy enough to use the null pointer passed along by JNA. I've modified the code to handle the null case, and will run our full test suite against the changes to check for breakage.

**Comment by antoinetran** [comment]
Ok, thank you! Just for the clarity, does "handle the null case" amounts to let  the null value pass or transforms it in some way? I would rather like the first option but I accept the second.

**Comment by lesserwhirls** [comment]
For the case of netCDF-4 (and HDF5), attributes with null values will be read in and stored as null, not transformed. When writing netCDF-4, attributes whose values have not been set will be seen by the C library as a null pointer, and thus stored as null (Dataspace Message 2 - null data space). Basically, once the PR is merged, you will be able to round trip null valued attributes in netCDF-4.

**Comment by lesserwhirls** [comment]
For a concrete example, check out this [test case](https://github.com/Unidata/thredds/pull/1262/files#diff-43f9cdcb162e8122ab6cfaba063bb7baR118).

**Comment by lesserwhirls** [comment]
Fixed by Unidata/thredds#1262. Thank you for working with us on this issue!
